### ISSUE TITLE
Improved version of code

### ISSUE THREAD

**Issue by chijames** [sfzhou5678/PolyEncoder#5]
Hi,

Thanks for sharing your code! However, there might be some potential issues in your implementation. To name a few:

1. We don't need linear, dropout, and normalize layer in the model.
2. The original paper does not use the softmax scaling trick. However, if it's removed, the performance will drop significantly using this repo's code. After some experiments, this is caused by 1. above.
3. Use the last hidden layer but not pooler_output of bert model is more reasonable and leads to better performance.
4. There are only m poly codes as we don't need the m+1 th for final aggregation. It's also not reasonable, since doing so expect that single vector to be used in every context-response pair.
5. The context representation is reversed on the level of sentence (transform.py ln 63), and the original paper uses newline (instead of [SEP]) to separate sentences, which may affect performance.
6. (minor) Learning rate schedule and gradient accumulation are not implemented correctly in the training loop.

Overall, fixing the above issues can resolve issue #1 and #4. As there are many modifications and some are non-trivial, I decide not to submit a pull request but adapt from your code and implement a new version. Here is the github repo link: https://github.com/chijames/Poly-Encoder. I try to follow the original paper as much as I can.

There are also some minor fixes and improvements. Specifically, add code for inference (this should resolve issue #3 ) and other evaluation metrics. The parsing and evaluation results of dstc 7 are also there. Feel free to check it out.

Thanks!

**Comment by sfzhou5678** [comment]
Ok thanks a lot! I have add your repo to my README, any further improvement would also be greatly appreciated!

**Comment by Wiselnn570** [comment]
Hi **@chijames** , Thanks for sharing! The solution #4 do work in my project with small samples.
